File: packages/google_mlkit_entity_extraction/android/src/main/java/com/google_mlkit_entity_extraction/EntityExtractor.java
package com.google_mlkit_entity_extraction;

import androidx.annotation.NonNull;

import com.google.mlkit.nl.entityextraction.DateTimeEntity;
import com.google.mlkit.nl.entityextraction.Entity;
import com.google.mlkit.nl.entityextraction.EntityAnnotation;
import com.google.mlkit.nl.entityextraction.EntityExtraction;
import com.google.mlkit.nl.entityextraction.EntityExtractionParams;
import com.google.mlkit.nl.entityextraction.EntityExtractionRemoteModel;
import com.google.mlkit.nl.entityextraction.EntityExtractorOptions;
import com.google.mlkit.nl.entityextraction.FlightNumberEntity;
import com.google.mlkit.nl.entityextraction.IbanEntity;
import com.google.mlkit.nl.entityextraction.IsbnEntity;
import com.google.mlkit.nl.entityextraction.MoneyEntity;
import com.google.mlkit.nl.entityextraction.PaymentCardEntity;
import com.google.mlkit.nl.entityextraction.TrackingNumberEntity;
import com.google_mlkit_commons.GenericModelManager;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.TimeZone;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class EntityExtractor implements MethodChannel.MethodCallHandler {
    private static final String START = "nlp#startEntityExtractor";
    private static final String CLOSE = "nlp#closeEntityExtractor";
    private static final String MANAGE = "nlp#manageEntityExtractionModels";

    private final Map<String, com.google.mlkit.nl.entityextraction.EntityExtractor> instances = new HashMap<>();
    private final GenericModelManager genericModelManager = new GenericModelManager();

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        String method = call.method;
        switch (method) {
            case START:
                extractEntities(call, result);
                break;
            case CLOSE:
                closeDetector(call);
                result.success(null);
                break;
            case MANAGE:
                manageModel(call, result);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private void extractEntities(MethodCall call, final MethodChannel.Result result) {
        String text = call.argument("text");

        String id = call.argument("id");
        com.google.mlkit.nl.entityextraction.EntityExtractor extractor = instances.get(id);
        if (extractor == null) {
            String language = call.argument("language");
            extractor = EntityExtraction.getClient(
                    new EntityExtractorOptions.Builder(language)
                            .build());
            instances.put(id, extractor);
        }
        final com.google.mlkit.nl.entityextraction.EntityExtractor entityExtractor = extractor;

        Map<String, Object> parameters = call.argument("parameters");
        Set<Integer> filters = null;
        if (parameters.get("filters") != null) {
            filters = new HashSet<>((List<Integer>) parameters.get("filters"));
        }

        Locale locale = null;
        if (parameters.get("locale") != null) {
            locale = new Locale.Builder().setLanguage((String) parameters.get("locale")).build();
        }

        TimeZone timeZone = null;
        if (parameters.get("timezone") != null) {
            timeZone = TimeZone.getTimeZone((String) parameters.get("timezone"));
        }

        Long referenceTime = null;
        if (parameters.get("time") != null) {
            referenceTime = (Long) parameters.get("time");
        }

        EntityExtractionParams params = new EntityExtractionParams.Builder(text)
                .setEntityTypesFilter(filters)
                .setPreferredLocale(locale)
                .setReferenceTimeZone(timeZone)
                .setReferenceTime(referenceTime)
                .build();

        entityExtractor
                .downloadModelIfNeeded()
                .addOnSuccessListener(
                        aVoid -> {
                            // Model downloading succeeded, you can call the extraction API here.
                            entityExtractor.annotate(params)
                                    .addOnSuccessListener(entityAnnotations -> {
                                        List<Map<String, Object>> allAnnotations = new ArrayList<>(entityAnnotations.size());

                                        for (EntityAnnotation entityAnnotation : entityAnnotations) {
                                            Map<String, Object> annotation = new HashMap<>();
                                            List<Entity> entities = entityAnnotation.getEntities();
                                            annotation.put("text", entityAnnotation.getAnnotatedText());
                                            annotation.put("start", entityAnnotation.getStart());
                                            annotation.put("end", entityAnnotation.getEnd());

                                            List<Map<String, Object>> allEntities = new ArrayList<>();
                                            for (Entity entity : entities) {
                                                Map<String, Object> entityData = new HashMap<>();
                                                entityData.put("type", entity.getType());
                                                entityData.put("raw", entity.toString());
                                                switch (entity.getType()) {
                                                    case Entity.TYPE_ADDRESS:
                                                    case Entity.TYPE_URL:
                                                    case Entity.TYPE_PHONE:
                                                    case Entity.TYPE_EMAIL:
                                                        break;
                                                    case Entity.TYPE_DATE_TIME:
                                                        DateTimeEntity dateTimeEntity = entity.asDateTimeEntity();
                                                        entityData.put("dateTimeGranularity", dateTimeEntity.getDateTimeGranularity() + 1);
                                                        entityData.put("timestamp", dateTimeEntity.getTimestampMillis());
                                                        break;
                                                    case Entity.TYPE_FLIGHT_NUMBER:
                                                        FlightNumberEntity flightNumberEntity = entity.asFlightNumberEntity();
                                                        entityData.put("code", flightNumberEntity.getAirlineCode());
                                                        entityData.put("number", flightNumberEntity.getFlightNumber());
                                                        break;
                                                    case Entity.TYPE_IBAN:
                                                        IbanEntity ibanEntity = entity.asIbanEntity();
                                                        entityData.put("iban", ibanEntity.getIban());
                                                        entityData.put("code", ibanEntity.getIbanCountryCode());
                                                        break;
                                                    case Entity.TYPE_ISBN:
                                                        IsbnEntity isbnEntity = entity.asIsbnEntity();
                                                        entityData.put("isbn", isbnEntity.getIsbn());
                                                        break;
                                                    case Entity.TYPE_MONEY:
                                                        MoneyEntity moneyEntity = entity.asMoneyEntity();
                                                        entityData.put("fraction", moneyEntity.getFractionalPart());
                                                        entityData.put("integer", moneyEntity.getIntegerPart());
                                                        entityData.put("unnormalized", moneyEntity.getUnnormalizedCurrency());
                                                        break;
                                                    case Entity.TYPE_PAYMENT_CARD:
                                                        PaymentCardEntity paymentCardEntity = entity.asPaymentCardEntity();
                                                        entityData.put("network", paymentCardEntity.getPaymentCardNetwork());
                                                        entityData.put("number", paymentCardEntity.getPaymentCardNumber());
                                                        break;
                                                    case Entity.TYPE_TRACKING_NUMBER:
                                                        TrackingNumberEntity trackingNumberEntity = entity.asTrackingNumberEntity();
                                                        entityData.put("carrier", trackingNumberEntity.getParcelCarrier());
                                                        entityData.put("number", trackingNumberEntity.getParcelTrackingNumber());
                                                        break;
                                                }

                                                allEntities.add(entityData);
                                            }
                                            annotation.put("entities", allEntities);
                                            allAnnotations.add(annotation);
                                        }

                                        result.success(allAnnotations);
                                    })
                                    .addOnFailureListener(e -> result.error("BarcodeDetectorError", e.toString(), null));
                        })
                .addOnFailureListener(
                        e -> {
                            // Model could not be downloaded or other internal error.
                            result.error("Error building extractor", "Model not downloaded", null);
                        });
    }

    private void closeDetector(MethodCall call) {
        String id = call.argument("id");
        com.google.mlkit.nl.entityextraction.EntityExtractor entityExtractor = instances.get(id);
        if (entityExtractor == null) return;
        entityExtractor.close();
        instances.remove(id);
    }

    private void manageModel(MethodCall call, final MethodChannel.Result result) {
        EntityExtractionRemoteModel model =
                new EntityExtractionRemoteModel.Builder(call.argument("model")).build();
        genericModelManager.manageModel(model, call, result);
    }
}

File: packages/google_mlkit_entity_extraction/android/src/main/kotlin/com/google_mlkit_entity_extraction/EntityExtractor.kt
package com.google_mlkit_entity_extraction

import com.google.mlkit.nl.entityextraction.Entity
import com.google.mlkit.nl.entityextraction.EntityExtraction
import com.google.mlkit.nl.entityextraction.EntityExtractionParams
import com.google.mlkit.nl.entityextraction.EntityExtractionRemoteModel
import com.google.mlkit.nl.entityextraction.EntityExtractorOptions
import com.google_mlkit_commons.GenericModelManager
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.Locale
import java.util.TimeZone

class EntityExtractor : MethodChannel.MethodCallHandler {
    companion object {
        private const val START = "nlp#startEntityExtractor"
        private const val CLOSE = "nlp#closeEntityExtractor"
        private const val MANAGE = "nlp#manageEntityExtractionModels"
    }

    private val instances = HashMap<String, com.google.mlkit.nl.entityextraction.EntityExtractor>()
    private val genericModelManager = GenericModelManager()

    override fun onMethodCall(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        when (call.method) {
            START -> {
                extractEntities(call, result)
            }

            CLOSE -> {
                closeDetector(call)
                result.success(null)
            }

            MANAGE -> {
                manageModel(call, result)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    private fun extractEntities(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val text = call.argument<String>("text")
        val id = call.argument<String>("id")

        var extractor = instances[id]
        if (extractor == null) {
            val language = call.argument<String>("language")!!
            extractor =
                EntityExtraction.getClient(
                    EntityExtractorOptions.Builder(language).build(),
                )
            instances[id!!] = extractor
        }

        val entityExtractor = extractor

        val parameters = call.argument<Map<String, Any>>("parameters")!!
        val filters = (parameters["filters"] as? List<Int>)?.toHashSet()

        val locale =
            (parameters["locale"] as? String)?.let {
                Locale.Builder().setLanguage(it).build()
            }

        val timeZone =
            (parameters["timezone"] as? String)?.let {
                TimeZone.getTimeZone(it)
            }

        val referenceTime = parameters["time"] as? Long

        val params =
            EntityExtractionParams
                .Builder(text!!)
                .setEntityTypesFilter(filters)
                .setPreferredLocale(locale)
                .setReferenceTimeZone(timeZone)
                .setReferenceTime(referenceTime)
                .build()

        entityExtractor
            .downloadModelIfNeeded()
            .addOnSuccessListener {
                entityExtractor
                    .annotate(params)
                    .addOnSuccessListener { entityAnnotations ->
                        val allAnnotation =
                            entityAnnotations.map { entityAnnotation ->
                                val allEntities =
                                    entityAnnotation.entities.map { entity ->
                                        val entityData = HashMap<String, Any>()
                                        entityData["type"] = entity.type
                                        entityData["raw"] = entity.toString()

                                        when (entity.type) {
                                            Entity.TYPE_ADDRESS,
                                            Entity.TYPE_URL,
                                            Entity.TYPE_PHONE,
                                            Entity.TYPE_EMAIL,
                                            -> {
                                                Unit
                                            }

                                            Entity.TYPE_DATE_TIME -> {
                                                val dateTimeEntity = entity.asDateTimeEntity()
                                                entityData["dateTimeGranularity"] = dateTimeEntity!!.dateTimeGranularity + 1
                                                entityData["timestamp"] = dateTimeEntity.timestampMillis
                                            }

                                            Entity.TYPE_FLIGHT_NUMBER -> {
                                                val flightNumberEntity = entity.asFlightNumberEntity()
                                                entityData["code"] = flightNumberEntity!!.airlineCode
                                                entityData["number"] = flightNumberEntity.flightNumber
                                            }

                                            Entity.TYPE_IBAN -> {
                                                val ibanEntity = entity.asIbanEntity()
                                                entityData["iban"] = ibanEntity!!.iban
                                                entityData["code"] = ibanEntity.ibanCountryCode
                                            }

                                            Entity.TYPE_ISBN -> {
                                                entityData["isbn"] = entity.asIsbnEntity()!!.isbn
                                            }

                                            Entity.TYPE_MONEY -> {
                                                val moneyEntity = entity.asMoneyEntity()
                                                entityData["fraction"] = moneyEntity!!.fractionalPart
                                                entityData["integer"] = moneyEntity.integerPart
                                                entityData["unnormalized"] = moneyEntity.unnormalizedCurrency
                                            }

                                            Entity.TYPE_PAYMENT_CARD -> {
                                                val paymentCardEntity = entity.asPaymentCardEntity()
                                                entityData["network"] = paymentCardEntity!!.paymentCardNetwork
                                                entityData["number"] = paymentCardEntity.paymentCardNumber
                                            }

                                            Entity.TYPE_TRACKING_NUMBER -> {
                                                val trackingNumberEntity = entity.asTrackingNumberEntity()
                                                entityData["carrier"] = trackingNumberEntity!!.parcelCarrier
                                                entityData["number"] = trackingNumberEntity.parcelTrackingNumber
                                            }
                                        }
                                        entityData
                                    }

                                hashMapOf(
                                    "text" to entityAnnotation.annotatedText,
                                    "start" to entityAnnotation.start,
                                    "end" to entityAnnotation.end,
                                    "entities" to allEntities,
                                )
                            }
                        result.success(allAnnotation)
                    }.addOnFailureListener { e ->
                        result.error("BarcodeDetectorError", e.toString(), null)
                    }
            }.addOnFailureListener {
                result.error("Error building extractor", "Model not downloaded", null)
            }
    }

    private fun closeDetector(call: MethodCall) {
        val id = call.argument<String>("id") ?: return
        val entityExtractor = instances[id] ?: return
        entityExtractor.close()
        instances.remove(id)
    }

    private fun manageModel(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val model = EntityExtractionRemoteModel.Builder(call.argument<String>("model")!!).build()
        genericModelManager.manageModel(model, call, result)
    }
}
